File: src/main/swift/g0001_0100/s0007_reverse_integer/Solution.swift
// #Medium #Top_Interview_Questions #Math #Udemy_Integers
// #2024_06_18_Time_0_ms_(100.00%)_Space_15.3_MB_(69.27%)

class Solution {
    func reverse(_ x: Int) -> Int {
        var rev: Int = 0
        var number = x

        while number != 0 {
            let pop = number % 10
            number /= 10

            if rev > Int32.max / 10 || (rev == Int32.max / 10 && pop > 7) {
                return 0
            }
            if rev < Int32.min / 10 || (rev == Int32.min / 10 && pop < -8) {
                return 0
            }
            rev = rev * 10 + pop
        }

        return rev
    }
}

File: src/main/rust/g0001_0100/s0007_reverse_integer/Solution.rs
// #Medium #Top_Interview_Questions #Math #Udemy_Integers
// #2024_09_04_Time_0_ms_(100.00%)_Space_2.1_MB_(18.54%)

impl Solution {
    pub fn reverse(x: i32) -> i32 {
        let mut x = x;
        let mut rev: i64 = 0; // Use i64 to handle overflow

        while x != 0 {
            rev = rev * 10 + (x % 10) as i64;
            x /= 10;
        }

        if rev > i32::MAX as i64 || rev < i32::MIN as i64 {
            return 0;
        }

        rev as i32
    }
}
